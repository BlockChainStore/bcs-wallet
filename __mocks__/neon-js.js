const neonjs = jest.genMockFromModule('neon-js')

const promiseMockGen = (result, error = false) => {
  return jest.fn(() => {
    return new Promise((resolve, reject) => {
      if (error) reject(error)
      resolve(result)
    })
  })
}

const privateKey = 'L4AJ14CNaBWPemRJKC34wyZwbmxg33GETs4Y1F8uK7rRmZ2UHrJn'
const address = 'AM22coFfbe9N6omgL9ucFBLkeaMNg9TEyL'
const encryptedKey = '6PYUGtvXiT5TBetgWf77QyAFidQj61V8FJeFBFtYttmsSxcbmP4vCFRCWu'
const scriptHash = '4bcdc110b6514312ead9420467475232d4f08539'

neonjs.api = {
  neonDB: {
    getClaims: jest.fn(),
    doClaimAllGas: promiseMockGen({ result: true }),
    doSendAsset: promiseMockGen({ result: true }),
    getWalletDBHeight: promiseMockGen(586435),
    getAPIEndpoint: jest.fn(() => 'http://testnet-api.wallet.cityofzion.io'),
    getRPCEndpoint: promiseMockGen(''),
    doMintTokens: promiseMockGen({ result: true }),
    getTransactionHistory: promiseMockGen([]),
    getBalance: promiseMockGen({ NEO: { balance: 1 }, GAS: { balance: 1 } })
  },
  nep5: {
    getTokenInfo: promiseMockGen({ result: true }),
    getTokenBalance: promiseMockGen(100)
  }
}

neonjs.create = {
  account: {
    address
    // privateKey
    // wif
  }
}
neonjs.tx = {
  serializeTransaction: jest.fn()
}
// TODO - look into why I chose to use encrypt vs encryptWif from new API
neonjs.wallet = {
  getPublicKeyEncoded: jest.fn(),
  decryptWIF: jest.fn((wif) => {
    return new Promise((resolve, reject) => {
      if (!wif) reject(new Error())
      resolve(privateKey)
    })
  }),
  decrypt: jest.fn(() => privateKey),
  encrypt: jest.fn(() => encryptedKey),
  generatePrivateKey: jest.fn(() => privateKey),
  Account: jest.fn(() => { return { address } }),
  getVerificationScriptFromPublicKey: jest.fn(() => scriptHash),
  isAddress: jest.fn(() => true),
  isNEP2: jest.fn(() => true)
}

module.exports = neonjs
