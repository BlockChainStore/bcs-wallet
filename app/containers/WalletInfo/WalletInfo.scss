@import '../../styles/variables';
.tokenBalances {
    padding-top: 10px;
}

.refreshIconContainer {
    padding-top: 30px; //
    position: absolute;
    top: 10px;
    left: 50%;
    transform: translate(-50%);
}

.refreshIcon {
    padding-top: 30px; //
    font-size: 2.2em;
    cursor: pointer;
    color: $default-button;
    &:hover {
        color: $default-button-hover;
    }
}