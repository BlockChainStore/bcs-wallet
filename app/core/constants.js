// @flow
export const NEON_WALLET_RELEASE_LINK =
  'https://github.com/CityOfZion/neon-wallet/releases'

export const ASSETS = {
  NEO: 'NEO',
  GAS: 'GAS'
}

export const EXPLORERS = {
  NEO_TRACKER: 'Neotracker',
  NEO_SCAN: 'Neoscan',
  ANT_CHAIN: 'Antchain',
  NEO_VERSE: 'NEOverse'
}

export const ROUTES = {
  HOME: '',
  DASHBOARD: '/dashboard',
  CREATE_WALLET: '/create',
  ENCRYPT_KEY: '/encrypt-key',
  LOGIN_PRIVATE_KEY: '/login-private-key',
  LOGIN_LOCAL_STORAGE: '/login-local-storage',
  LOGIN_LEDGER_NANO_S: '/login-ledger-nano-s',
  LOGIN_NEP2: '/login-encrypted',
  LOGIN_TOKEN_SALE: '/login-token-sale',
  TOKEN_SALE: '/token-sale',
  SETTINGS: '/settings',
  DISPLAY_WALLET_KEYS: '/display-wallet-keys'
}

export const NOTIFICATION_LEVELS = {
  ERROR: 'error',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning'
}

export const NOTIFICATION_POSITIONS = {
  TOP_CENTER: 'tc',
  TOP_RIGHT: 'tr',
  TOP_LEFT: 'tl',
  BOTTOM_CENTER: 'bc',
  BOTTOM_RIGHT: 'br',
  BOTTOM_LEFT: 'bl'
}

export const BIP44_PATH =
  '8000002C' + '80000378' + '80000000' + '00000000' + '00000000'

export const MODAL_TYPES = {
  SEND: 'SEND',
  RECEIVE: 'RECEIVE',
  CONFIRM: 'CONFIRM',
  TOKEN_INFO: 'TOKEN_INFO',
  TOKEN: 'TOKEN',
  ICO: 'ICO'
}

export const MAIN_NETWORK_ID = '1'
export const TEST_NETWORK_ID = '2'

// TestNet
export const TOKENS_TEST = {
  BCS : 'e81c837d37763afbf894d319c868a09f219f2be9'
}

// MainNet
export const TOKENS = {
  BCS : '546c5872a992b2754ef327154f4c119baabff65f'
}

export const DEFAULT_WALLET = {
  name: 'userWallet',
  version: '1.0',
  scrypt: {
    cost: 16384,
    blockSize: 8,
    parallel: 8,
    size: 64
  },
  accounts: [],
  extra: null
}

export const DEFAULT_CURRENCY_CODE = 'usd'

export const CURRENCIES = {
  aud: { symbol: '$' },
  brl: { symbol: 'R$' },
  cad: { symbol: '$' },
  chf: { symbol: 'Fr.' },
  clp: { symbol: '$' },
  cny: { symbol: '¥' },
  czk: { symbol: 'Kč' },
  dkk: { symbol: 'kr. ' },
  eur: { symbol: '€' },
  gbp: { symbol: '£' },
  hkd: { symbol: '$' },
  huf: { symbol: 'Ft ' },
  idr: { symbol: 'Rp ' },
  ils: { symbol: '₪' },
  inr: { symbol: '₹' },
  jpy: { symbol: '¥' },
  krw: { symbol: '₩' },
  mxn: { symbol: '$' },
  myr: { symbol: 'RM' },
  nok: { symbol: 'kr ' },
  nzd: { symbol: '$' },
  php: { symbol: '₱' },
  pkr: { symbol: '₨ ' },
  pln: { symbol: 'zł' },
  rub: { symbol: '₽' },
  sek: { symbol: 'kr ' },
  sgd: { symbol: 'S$' },
  thb: { symbol: '฿' },
  try: { symbol: '₺' },
  twd: { symbol: 'NT$' },
  usd: { symbol: '$' },
  zar: { symbol: 'R ' }
}
