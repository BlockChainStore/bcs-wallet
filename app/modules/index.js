// @flow
import { combineReducers } from 'redux'
import api from './api'
import generateWallet from './generateWallet'
import claim from './claim'
import notifications from './notifications'
import modal from './modal'
import addressBook from './addressBook'

export default combineReducers({
  api,
  generateWallet,
  claim,
  notifications,
  modal,
  addressBook
})
