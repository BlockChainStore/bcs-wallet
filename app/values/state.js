// @flow
export type ProgressState = 'INITIAL' | 'LOADING' | 'LOADED' | 'FAILED'

export const INITIAL: ProgressState = 'INITIAL'
export const LOADING: ProgressState = 'LOADING'
export const LOADED: ProgressState = 'LOADED'
export const FAILED: ProgressState = 'FAILED'
