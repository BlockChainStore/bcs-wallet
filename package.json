{
    "name": "BCSwallet",
    "version": "0.1.1",
    "main": "./main.js",
    "description": "BCSwallet folked from neon-wallet, Light wallet for NEO blockchain",
    "homepage": "https://github.com/CityOfZion/neon-wallet",
    "author": "Ethan Fast <ejhfast@gmail.com> (https://github.com/Ejhfast)",
    "scripts": {
        "postinstall": "electron-rebuild --force",
        "start": "cross-env NODE_ENV=production electron .",
        "assets": "cross-env NODE_ENV=production webpack --config ./config/webpack.config.prod",
        "dev": "cross-env START_HOT=1 NODE_ENV=development node_modules/.bin/webpack-dev-server --config ./config/webpack.config.dev",
        "start-dev": "cross-env NODE_ENV=development electron . --enable-logging --remote-debugging-port=9222",
        "assets-dev": "cross-env NODE_ENV=development webpack --config ./config/webpack.config.dev",
        "test": "npm rebuild node-hid && jest --coverage",
        "test-ci": "yarn test && cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js",
        "test:e2e": "ava __e2e__/*.e2e.js --timeout=120s",
        "test-watch": "jest --watch",
        "pack": "build --dir",
        "dist": "yarn assets && yarn build",
        "lint": "eslint -c .eslintrc ./app --ext .js,.jsx --ignore-pattern dist/",
        "lint-fix": "eslint -c .eslintrc ./app --ext .js,.jsx --fix --ignore-pattern dist/",
        "prepush": "echo 'running flow...' && yarn flow && echo 'running lint...' && yarn lint && echo 'running tests...' && yarn test",
        "electron-rebuild": "electron-rebuild --force --module_dir . -w node-hid",
        "flow": "flow check",
        "release-master": "./scripts/release-master.sh",
        "clean": "rm -r app/dist/*"
    },
    "keywords": [],
    "license": "ISC",
    "build": {
        "appId": "com.electron.neon",
        "files": [
            "./main.js",
            "app/dist/index.html",
            "app/dist/bundle.js",
            "app/dist/style.css"
        ],
        "dmg": {
            "contents": [{
                    "x": 110,
                    "y": 150
                },
                {
                    "x": 240,
                    "y": 150,
                    "type": "link",
                    "path": "/Applications"
                }
            ]
        },
        "linux": {
            "target": [
                "AppImage",
                "deb"
            ]
        },
        "win": {
            "target": "nsis",
            "icon": "build/icon.ico"
        }
    },
    "dependencies": {
        "axios": "0.17.0",
        "babel": "^6.23.0",
        "base-x": "3.0.2",
        "bigi": "1.4.2",
        "bignumber.js": "5.0.0",
        "bs58check": "2.0.2",
        "buffer": "5.0.8",
        "classnames": "2.2.5",
        "cleave.js": "1.0.7",
        "coveralls": "3.0.0",
        "crypto-js": "3.1.9-1",
        "ecurve": "1.0.5",
        "electron-context-menu": "0.9.1",
        "electron-json-storage": "4.0.2",
        "electron-save-file": "1.0.2",
        "elliptic": "6.4.0",
        "es6-promisify": "^5.0.0",
        "file-loader": "1.1.5",
        "fs": "0.0.1-security",
        "global": "4.3.2",
        "history": "4.7.2",
        "isomorphic-fetch": "2.2.1",
        "js-scrypt": "0.2.0",
        "lodash": "4.17.4",
        "neon-js": "git+https://github.com/cityofzion/neon-js.git#2.3.1",
        "node-gyp": "^3.6.2",
        "node-hid": "file:node-hid",
        "prop-types": "^15.6.0",
        "qrcode": "0.9.0",
        "raf": "3.4.0",
        "react": "16.1.1",
        "react-click-outside": "3.0.0",
        "react-data-grid": "2.0.73",
        "react-dom": "16.1.1",
        "react-hot-loader": "3.1.1",
        "react-icons": "2.2.7",
        "react-modal": "3.1.2",
        "react-notification-system": "0.2.16",
        "react-redux": "5.0.6",
        "react-router-dom": "4.2.2",
        "react-split-pane": "0.1.68",
        "react-syntax-highlighter": "5.8.0",
        "react-test-renderer": "16.1.1",
        "react-tippy": "1.2.2",
        "recompose": "^0.26.0",
        "redux": "3.7.2",
        "redux-logger": "3.0.6",
        "redux-saga": "^0.16.0",
        "redux-thunk": "2.2.0",
        "reselect": "^3.0.1",
        "secure-random": "1.1.1",
        "sifter": "0.5.3",
        "spawn-rx": "^2.0.12",
        "uuid": "^3.2.1",
        "wif": "2.0.6"
    },
    "devDependencies": {
        "autoprefixer": "7.1.6",
        "ava": "0.23.0",
        "axios-mock-adapter": "1.9.0",
        "babel-cli": "6.26.0",
        "babel-core": "6.26.0",
        "babel-eslint": "8.0.2",
        "babel-loader": "7.1.2",
        "babel-plugin-lodash": "3.2.11",
        "babel-plugin-transform-async-to-generator": "6.24.1",
        "babel-polyfill": "6.26.0",
        "babel-preset-env": "1.6.1",
        "babel-preset-flow": "6.23.0",
        "babel-preset-react": "6.24.1",
        "babel-preset-stage-0": "6.24.1",
        "css-loader": "0.28.7",
        "electron": "1.7.9",
        "electron-builder": "19.49.4",
        "electron-devtools-installer": "2.2.1",
        "electron-packager": "9.1.0",
        "electron-rebuild": "1.6.0",
        "enzyme": "3.1.1",
        "enzyme-adapter-react-16": "1.1.0",
        "enzyme-to-json": "3.2.1",
        "eslint": "^4.18.1",
        "eslint-config-standard": "10.2.1",
        "eslint-config-standard-react": "5.0.0",
        "eslint-loader": "1.9.0",
        "eslint-plugin-flowtype": "2.39.1",
        "eslint-plugin-import": "2.8.0",
        "eslint-plugin-jest": "21.2.0",
        "eslint-plugin-jsx-a11y": "6.0.2",
        "eslint-plugin-node": "5.2.1",
        "eslint-plugin-promise": "3.6.0",
        "eslint-plugin-react": "7.4.0",
        "eslint-plugin-standard": "3.0.1",
        "extract-text-webpack-plugin": "3.0.2",
        "flow-bin": "0.58.0",
        "html-webpack-plugin": "2.30.1",
        "husky": "0.14.3",
        "identity-obj-proxy": "3.0.0",
        "jest": "21.2.1",
        "json-loader": "0.5.7",
        "node-sass": "4.6.0",
        "redux-devtools-extension": "2.13.2",
        "redux-mock-store": "1.3.0",
        "resolve-url-loader": "2.2.0",
        "rimraf": "2.6.2",
        "sass-loader": "6.0.6",
        "spectron": "3.7.2",
        "style-loader": "0.19.0",
        "uglifyjs-webpack-plugin": "1.0.1",
        "url-loader": "0.6.2",
        "webpack": "3.8.1",
        "webpack-dev-server": "2.9.4"
    },
    "jest": {
        "moduleNameMapper": {
            "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.js",
            "\\.(css|less|sass|scss)$": "identity-obj-proxy"
        },
        "testPathIgnorePatterns": [
            "__tests__/testHelpers.js",
            "__tests__/setupTests.js"
        ],
        "unmockedModulePathPatterns": [
            "node_modules/react/",
            "node_modules/enzyme/"
        ],
        "snapshotSerializers": [
            "enzyme-to-json/serializer"
        ],
        "setupTestFrameworkScriptFile": "<rootDir>/__tests__/setupTests.js"
    }
}