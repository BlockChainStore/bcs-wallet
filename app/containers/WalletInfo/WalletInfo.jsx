// @flow
import React, { Component } from 'react'
import classNames from 'classnames'
import { isNil, keyBy } from 'lodash'

import Claim from '../Claim'

import Tooltip from '../../components/Tooltip'
import Button from '../../components/Button'

import { formatGAS, formatFiat, formatNEO } from '../../core/formatters'
import { ASSETS, CURRENCIES, MODAL_TYPES } from '../../core/constants'
import { toNumber } from '../../core/math'

import MdSync from 'react-icons/lib/md/sync'

import styles from './WalletInfo.scss'

import TokenBalances from './TokenBalances'

type Props = {
  address: string,
  NEO: string,
  GAS: string,
  neoPrice: number,
  gasPrice: number,
  tokenBalances: Array<TokenBalanceType>,
  loadWalletData: Function,
  currencyCode: string,
  showModal: Function,
  oldParticipateInSale: Function,
  participateInSale: Function,
  allTokens: Array<TokenItemType>,
  setUserGeneratedTokens: Function,
  networks: Array<NetworkItemType>,
  networkId: string,
}

export default class WalletInfo extends Component<Props> {
  render () {
    const {
      address,
      NEO,
      GAS,
      neoPrice,
      gasPrice,
      tokenBalances,
      showModal,
      currencyCode,
      participateInSale,
      oldParticipateInSale,
      networks,
      networkId,
      allTokens,
      setUserGeneratedTokens,
      loadWalletData
    } = this.props

    if (isNil(address)) {
      return null
    }

    let neoValue = neoPrice && NEO !== '0' ? neoPrice * toNumber(NEO) : 0
    let gasValue = gasPrice && GAS !== '0' ? gasPrice * toNumber(GAS) : 0

    let totalValue = neoValue + gasValue

    const displayCurrencyCode = currencyCode.toUpperCase()
    const currencySymbol = CURRENCIES[currencyCode].symbol

    return (
      <div id='accountInfo'>
        <div id='balance'>
          <div className='split'>
            <div className='label'>{ASSETS.NEO}</div>
            <div className='amountBig amountNeo'>{formatNEO(NEO)}</div>
            <div className='fiat neoWalletValue'>
              {currencySymbol}
              {formatFiat(neoValue)} {displayCurrencyCode}
            </div>
          </div>
          <div className='split'>
            <div className='label'>{ASSETS.GAS}</div>
            <div className='amountBig amountGas'>
              <Tooltip title={formatGAS(GAS)} disabled={GAS === 0}>
                {formatGAS(GAS, true)}
              </Tooltip>
            </div>
            <div className='fiat gasWalletValue'>
              {currencySymbol}
              {formatFiat(gasValue)} {displayCurrencyCode}
            </div>
          </div>
          <div className='fiat walletTotal'>
            Total {currencySymbol}
            {formatFiat(totalValue)} {displayCurrencyCode}
          </div>
          <div
            onClick={loadWalletData}
            className={classNames(
              styles.refreshIconContainer,
              'refreshBalance'
            )}
          >
            <Tooltip title='Refresh account balance'>
              <MdSync id='refresh' className={styles.refreshIcon} />
            </Tooltip>
          </div>
        </div>
        <div className='spacer' />
        <Claim />
        <TokenBalances tokenBalances={tokenBalances} showModal={showModal} />
        <div
          className={(styles.walletButton, styles.icoButton)}
          onClick={() =>
            showModal(MODAL_TYPES.ICO, {
              assetBalances: {
                [ASSETS.NEO]: NEO,
                [ASSETS.GAS]: GAS
              },
              tokenBalances: keyBy(tokenBalances, 'symbol'),
              showTokensModal: () => {
                showModal(MODAL_TYPES.TOKEN, {
                  tokens: allTokens,
                  networks,
                  networkId,
                  setUserGeneratedTokens,
                  onSave: loadWalletData
                })
              },
              oldParticipateInSale,
              participateInSale
            })
          }
        >
            <Button>Participate in a token sale</Button>
        </div>
      </div>
    )
  }
}
