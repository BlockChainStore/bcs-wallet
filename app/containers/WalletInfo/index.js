// @flow
import { connect, type MapStateToProps } from 'react-redux'
import { bindActionCreators } from 'redux'
import { compose } from 'recompose'
import { values, omit } from 'lodash'

import accountActions from '../../actions/accountActions'
import pricesActions from '../../actions/pricesActions'
import balancesActions from '../../actions/balancesActions'
import withData from '../../hocs/api/withData'
import withActions from '../../hocs/api/withActions'
import withNetworkData from '../../hocs/withNetworkData'
import withAuthData from '../../hocs/withAuthData'
import withCurrencyData from '../../hocs/withCurrencyData'
import withFilteredTokensData from '../../hocs/withFilteredTokensData'
import withSuccessNotification from '../../hocs/withSuccessNotification'
import withFailureNotification from '../../hocs/withFailureNotification'
import { updateSettingsActions } from '../../actions/settingsActions'
import { getNetworks } from '../../core/networks'
import { showModal } from '../../modules/modal'
import { participateInSale, oldParticipateInSale } from '../../modules/sale'

import WalletInfo from './WalletInfo'

const mapStateToProps: MapStateToProps<*, *, *> = (state: Object) => ({
  networks: getNetworks()
})

const mapBalanceDataToProps = (balances) => ({
  NEO: balances.NEO,
  GAS: balances.GAS,
  tokenBalances: values(omit(balances, 'NEO', 'GAS'))
})

const mapPricesDataToProps = ({ NEO, GAS }) => ({
  neoPrice: NEO,
  gasPrice: GAS
})

const actionCreators = {
  showModal,
  participateInSale,
  oldParticipateInSale
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

const mapSettingsActionsToProps = (actions) => ({
  setUserGeneratedTokens: (tokens) => actions.request({ tokens })
})

const mapAccountActionsToProps = (actions, props) => ({
  loadWalletData: () => actions.request({ net: props.net, address: props.address, tokens: props.tokens })
})

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withData(pricesActions, mapPricesDataToProps),
  withData(balancesActions, mapBalanceDataToProps),
  withNetworkData(),
  withAuthData(),
  withCurrencyData('currencyCode'),
  withFilteredTokensData(),
  withActions(updateSettingsActions, mapSettingsActionsToProps),
  withActions(accountActions, mapAccountActionsToProps),
  withSuccessNotification(accountActions, 'Received latest blockchain information.'),
  withFailureNotification(accountActions, 'Failed to retrieve blockchain information.')
)(WalletInfo)
