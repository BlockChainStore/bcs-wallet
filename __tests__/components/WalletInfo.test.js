import React from 'react'
import * as neonjs from 'neon-js'
import { merge } from 'lodash'
import { mount, shallow } from 'enzyme'

import { createStore, provideStore, provideState } from '../testHelpers'
import WalletInfo from '../../app/containers/WalletInfo'
import { SHOW_NOTIFICATION } from '../../app/modules/notifications'
import { NOTIFICATION_LEVELS, DEFAULT_CURRENCY_CODE, MAIN_NETWORK_ID } from '../../app/core/constants'
import { LOADED } from '../../app/values/state'

// TODO research how to move the axios mock code which is repeated in NetworkSwitch to a helper or config file
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { version } from '../../package.json'

const axiosMock = new MockAdapter(axios)
axiosMock
  .onGet('http://testnet-api.wallet.cityofzion.io/v2/version')
  .reply(200, { version })
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/?limit=0&convert=USD')
  .reply(200, [{ symbol: 'NEO', price_usd: 24.5 }, { symbol: 'GAS', price_usd: 18.2 }])

jest.mock('electron', () => ({
  app: {
    getPath: () => 'C:\\tmp\\mock_path'
  }
}))
jest.useFakeTimers()

jest.unmock('qrcode')
import QRCode from 'qrcode/lib/browser' // eslint-disable-line
QRCode.toCanvas = jest.fn()

const initialState = {
  api: {
    NETWORK: {
      batch: false,
      state: LOADED,
      data: MAIN_NETWORK_ID
    },
    AUTH: {
      batch: false,
      state: LOADED,
      data: {
        address: 'ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw'
      }
    },
    SETTINGS: {
      batch: false,
      state: LOADED,
      data: {
        currency: DEFAULT_CURRENCY_CODE,
        tokens: []
      }
    },
    PRICES: {
      batch: false,
      state: LOADED,
      data: {
        NEO: 25.48,
        GAS: 18.1
      }
    },
    BALANCES: {
      batch: false,
      state: LOADED,
      data: {
        NEO: '100001',
        GAS: '1000.0001601'
      }
    },
    CLAIMS: {
      batch: false,
      state: LOADED,
      data: {
        total: '0.5'
      }
    }
  },
  claim: {
    claimRequest: false,
    disableClaimButton: false
  }
}

describe('WalletInfo', () => {
  test('renders without crashing', () => {
    const store = createStore(initialState)
    const wrapper = shallow(<WalletInfo store={store} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('correctly renders data from state', () => {
    const wrapper = mount(provideState(<WalletInfo />, initialState))

    const neoWalletValue = wrapper.find('.neoWalletValue')
    const gasWalletValue = wrapper.find('.gasWalletValue')
    const walletValue = wrapper.find('.walletTotal')

    const expectedNeoWalletValue = '2,548,025.48'
    const expectedGasWalletValue = '18,100.00'
    const expectedWalletValue = '2,566,125.48'
    const neoField = wrapper.find('.amountNeo')
    const gasField = wrapper.find('.amountGas')

    expect(neoWalletValue.text()).toEqual(`$${expectedNeoWalletValue} USD`)
    expect(gasWalletValue.text()).toEqual(`$${expectedGasWalletValue} USD`)
    expect(walletValue.text()).toEqual(`Total $${expectedWalletValue} USD`)
    expect(neoField.text()).toEqual('100,001')
    // TODO: Test the GAS tooltip value, this is testing the display value, truncated to 4 decimals
    expect(gasField.text()).toEqual('1,000.0002')
  })

  test('account data refreshes when refresh button is clicked', () => {
    const store = createStore(initialState)
    const wrapper = mount(provideStore(<WalletInfo />, store))

    wrapper.find('.refreshBalance').simulate('click')

    expect(store.getActions()).toContainEqual(expect.objectContaining({
      type: 'BATCH/REQUEST',
      meta: expect.objectContaining({ id: 'ACCOUNT' })
    }))
  })

  test('correctly renders data from state with non-default currency', () => {
    const testState = merge(initialState, {
      api: {
        SETTINGS: { data: { currency: 'eur' } },
        PRICES: { data: { NEO: 1.11, GAS: 0.55 } }
      }
    })
    const wrapper = mount(provideState(<WalletInfo />, testState))

    const neoWalletValue = wrapper.find('.neoWalletValue')
    const gasWalletValue = wrapper.find('.gasWalletValue')
    const walletValue = wrapper.find('.walletTotal')

    const expectedNeoWalletValue = '111,001.11'
    const expectedGasWalletValue = '550.00'
    const expectedWalletValue = '111,551.11'

    expect(neoWalletValue.text()).toEqual(`€${expectedNeoWalletValue} EUR`)
    expect(gasWalletValue.text()).toEqual(`€${expectedGasWalletValue} EUR`)
    expect(walletValue.text()).toEqual(`Total €${expectedWalletValue} EUR`)
  })
})
