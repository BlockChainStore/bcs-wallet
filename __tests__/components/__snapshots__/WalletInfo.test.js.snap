// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletInfo renders without crashing 1`] = `
<Connect(withData(Connect(withData(Connect(withData(withProps(Connect(withData(Connect(withData(Connect(withData(withProps(Connect(withData(Connect(withActions(Connect(withActions(Connect(Connect(withProgressProp(LoadedNotifier)))))))))))))))))))))))
  networks={
    Array [
      Object {
        "id": "1",
        "label": "MainNet",
        "network": "MainNet",
      },
      Object {
        "id": "2",
        "label": "TestNet",
        "network": "TestNet",
      },
    ]
  }
  oldParticipateInSale={[Function]}
  participateInSale={[Function]}
  showModal={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
